### ISSUE TITLE
Issues found testing P4P 4.0.0 pre-release

### ISSUE THREAD

**Issue by mdavidsaver** [mdavidsaver/p4p#76]
<div class="moz-cite-prefix">On 9/9/21 11:31 AM, Chandler, Elaine wrote:<br>
</div>
<blockquote type="cite" cite="mid:DM8PR09MB7128D68B332440927F28030EB7D59@DM8PR09MB7128.namprd09.prod.outlook.com">

  <span>Hi Michael,</span>
  <div><br>
  </div>

  <div>We ran some load tests on the PV Gateway on the P4P pre-release 
(4.0) and found a couple potential issues that we wanted to make sure 
you were aware of.</div>

  <div><br>
  </div>

  <div>
<div>
<div><span><b>1. PV clients not receiving data when data connection is saturated</b><br>
</span></div>
</div>
  </div>

  <div>We tested increasing number of PV monitors and found when the 
data rate is more than the NIC data limit, the PV monitors weren't 
receiving any data.  In P2P and the P4P 3.5, it would still be sending 
data, but at a lower data rate. Below shows the receiving
 sample and data rates of the PV monitors.  <br>
  </div>

  <div><br>
  </div>

  <div>

Gateway | # monitors | avg sample rate (Hz) | avg data rate (MiB/s) | total data rate (MB/s)
-- | -- | -- | -- | --
p4p 4.0 | 2 | 10.00 | 2370 | 4740
p4p 4.0 | 3 | 10.00 | 2370 | 7110
p4p 4.0 | 4 | 9.74 | 2308 | 9232
p4p 4.0 | 5 | 9.70 | 2299 | 11495
p4p 4.0 | 6 | 0.03 | 8 | 48
p4p 4.0 | 7 | 0.00 | 0 | 0
p4p 4.0 | 8 | 0.00 | 0 | 0
p4p 4.0 | 9 | 0.00 | 0 | 0
p4p 4.0 | 10 | 0.11 | 26 | 263
p4p 4.0 | 11 | 0.07 | 15 | 169
p4p 4.0 | 12 | 0.00 | 0 | 0

</div>
  </div>

  <div><br>
  </div>

  <div><br>
  </div>

  <div><b>2. PV Gateway crashes when running with machine with more load</b><br>
  </div>

  <div><br>
  </div>

  <div>We found that when we ran PV Gateway with our regular load test, 
it would crash with either a "corrupt double-link list" or "free(): 
corrupted unsorted chunks" error.  In our test scenario we have 3 IOCs, 
the PV Gateway, and 3 other services running on
 the server.  I noticed that the memory usage for the gateway goes up 
right before the crash:</div>
</blockquote>

<img width="776" alt="image" src="https://user-images.githubusercontent.com/4429215/137781797-b54e6c4b-53bc-42fa-aa69-4e407a04ae03.png">

**Comment by mdavidsaver** [comment]
I need more information as I'm unable to replicate either a stall or a crash.

**Comment by echandler-anl** [comment]
The memory is the resident memory from top.

I attached the the profile from perf running the test with 8 PV monitors.

[p4pperf.zip](https://github.com/mdavidsaver/p4p/files/7522318/p4pperf.zip)

 The .out file is from running `perf report -n --stdio` and the .svg is a [framegraph](https://github.com/brendangregg/FlameGraph) from the profile.   I noticed that for one of the runs it also crashed with the same "corrupt double-link list" error, where the memory usage went up to 20GB.
